{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": true,
    "editable": true
   },
   "source": [
    "# Create input files for simulations based on experimental fits\n",
    "\n",
    "<div id=\"toc-wrapper\"><h3> Table of Contents </h3><div id=\"toc\" style=\"max-height: 787px;\"><ol class=\"toc-item\"><li><a href=\"#Globals\">Globals</a></li><li><a href=\"#Create-stall-strength-file-for-Run-3-initiation-mutant-prediction-based-on-fit-of-Run-2-to-single-mutant-data-in-3-different-models-(for-fig.-4A,-fig.-4-supplement-1A-G)\">Create stall strength file for Run 3 initiation mutant prediction based on fit of Run 2 to single mutant data in 3 different models (for fig. 4A, fig. 4 supplement 1A-G)</a></li><li><a href=\"#Create-mRNA-sequence-files-for-Run-3-to-predict-initiation-mutant-YFP-synthesis-rates-in-3-different-models--(for-fig.-4A,-fig.-4-supplement-1A-G)\">Create mRNA sequence files for Run 3 to predict initiation mutant YFP synthesis rates in 3 different models  (for fig. 4A, fig. 4 supplement 1A-G)</a></li><li><a href=\"#Create-stall-strength-file-for-Run-14-serine-initiation-mutant-prediction-based-on-fit-of-Run-13-to-single-mutant-data-in-3-different-models--(for-fig.-4-supplement-1H)\">Create stall strength file for Run 14 serine initiation mutant prediction based on fit of Run 13 to single mutant data in 3 different models  (for fig. 4 supplement 1H)</a></li><li><a href=\"#Create-mRNA-sequence-files-for-Run-14-to-predict-serine-initiation-mutant-YFP-synthesis-rates-in-3-different-models-(for-fig.-4-supplement-1H)\">Create mRNA sequence files for Run 14 to predict serine initiation mutant YFP synthesis rates in 3 different models (for fig. 4 supplement 1H)</a></li><li><a href=\"#Create-stall-strength-file-for-Run-4-CTC,-CTT-double-mutant-prediction-based-on-fit-of-Run-2-to-single-mutant-data-(for-fig.-5-supplement-1A-B)\">Create stall strength file for Run 4 CTC, CTT double mutant prediction based on fit of Run 2 to single mutant data (for fig. 5 supplement 1A-B)</a></li><li><a href=\"#Create-mRNA-sequence-files-for-Run-4-to-predict-CTC-CTT-double-mutant-YFP-synthesis-rates-in-different-models-(for-fig.-5-supplement-1A-B)\">Create mRNA sequence files for Run 4 to predict CTC CTT double mutant YFP synthesis rates in different models (for fig. 5 supplement 1A-B)</a></li><li><a href=\"#Create-stall-strength-file-for-Run-16-CTA-multiple-mutant-prediction-based-on-fit-of-Run-2-to-single-mutant-data-(for-fig.-5)\">Create stall strength file for Run 16 CTA multiple mutant prediction based on fit of Run 2 to single mutant data (for fig. 5)</a></li><li><a href=\"#Create-mRNA-sequence-files-for-Run-16-to-predict-multiple-mutant-YFP-synthesis-rates-in-different-models-(for-fig.-5)\">Create mRNA sequence files for Run 16 to predict multiple mutant YFP synthesis rates in different models (for fig. 5)</a></li><li><a href=\"#Create-stall-strength-file-for-Run-15-serine-double-mutant-prediction-based-on-fit-of-Run-13-to-single-mutant-data-(for-fig.-5-supplement-1C)\">Create stall strength file for Run 15 serine double mutant prediction based on fit of Run 13 to single mutant data (for fig. 5 supplement 1C)</a></li><li><a href=\"#Create-mRNA-sequence-files-for-Run-15-to-predict--serine-double-mutant-YFP-synthesis-rates-in-different-models-(for-fig.-5-supplement-1C)\">Create mRNA sequence files for Run 15 to predict  serine double mutant YFP synthesis rates in different models (for fig. 5 supplement 1C)</a></li><li><a href=\"#Create-stall-strength-file-for-Run-5-CTC-distance-mutant-prediction-based-on-fit-of-Run-2-to-single-mutant-data-(for-fig.-6-supplement-1)\">Create stall strength file for Run 5 CTC distance mutant prediction based on fit of Run 2 to single mutant data (for fig. 6 supplement 1)</a></li><li><a href=\"#Create-mRNA-sequence-files-for-Run-5-to-predict-CTC-distance-mutant-YFP-synthesis-rates-in-different-models-(for-fig.-6-supplement-1)\">Create mRNA sequence files for Run 5 to predict CTC distance mutant YFP synthesis rates in different models (for fig. 6 supplement 1)</a></li><li><a href=\"#Create-stall-strength-file-for-Runs-6,7,8-to-systematically-vary-initiation-rate,-number-of-stall-sites-and-distance-between-stall-sites-in-different-models-(for-Fig.-3)\">Create stall strength file for Runs 6,7,8 to systematically vary initiation rate, number of stall sites and distance between stall sites in different models (for Fig. 3)</a></li><li><a href=\"#Create-mRNA-sequence-files-for-Run-6-to-vary-initiation-rate-in-different-models-(for-Fig.-3A)\">Create mRNA sequence files for Run 6 to vary initiation rate in different models (for Fig. 3A)</a></li><li><a href=\"#Create-mRNA-sequence-files-for-Run-7-to-vary-number-of-stall-sites-in-different-models-(for-Fig.-3B)\">Create mRNA sequence files for Run 7 to vary number of stall sites in different models (for Fig. 3B)</a></li><li><a href=\"#Create-mRNA-sequence-files-for-Run-8-to-vary-distance-between-stall-sites-in-different-models-(for-Fig.-3C)\">Create mRNA sequence files for Run 8 to vary distance between stall sites in different models (for Fig. 3C)</a></li><li><a href=\"#Create-stall-strength-file-for-Run-11-CTA-distance-mutant-prediction-based-on-fit-of-Run-2-to-single-mutant-data-(for-Fig.-6)\">Create stall strength file for Run 11 CTA distance mutant prediction based on fit of Run 2 to single mutant data (for Fig. 6)</a></li><li><a href=\"#Create-mRNA-sequence-files-for-Run-11-to-predict-CTA-distance-mutant-YFP-synthesis-rates-in-different-models-(for-Fig.-6)\">Create mRNA sequence files for Run 11 to predict CTA distance mutant YFP synthesis rates in different models (for Fig. 6)</a></li></ol></div></div>\n",
    "\n",
    "These input files are for simulation runs 3 â€“ 16."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": true,
    "editable": true
   },
   "source": [
    "### Globals"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# os commands\n",
    "import os\n",
    "# sequence input and output\n",
    "import Bio.SeqIO\n",
    "# provides dictionary of codon names\n",
    "from Bio.SeqUtils.CodonUsage import SynonymousCodons\n",
    "# for converting 3 letter amino acid code to 1 letter code\n",
    "from Bio.SeqUtils import seq1\n",
    "# for tab data processing\n",
    "import pandas as pd\n",
    "# numeric and matrix library\n",
    "import numpy as np\n",
    "# shell utilities\n",
    "import shutil\n",
    "# for submitting shell commands\n",
    "import subprocess as sp\n",
    "\n",
    "# create a dictionary of codon names and number (arranged alphabetically by aa)\n",
    "codonnum = 0\n",
    "codonDict = dict()\n",
    "for aa in sorted(SynonymousCodons, key=lambda aa3: seq1(aa3)):\n",
    "    if aa == 'STOP':\n",
    "        continue\n",
    "    for codon in sorted(SynonymousCodons[aa]):\n",
    "        # these two codons are numbered out of order for consistent notation\n",
    "        # with Subramaniam et al. Cell 2014\n",
    "        if codon in ['AGC']:\n",
    "            codonDict['AGC'] = 59\n",
    "        elif codon in ['AGT']:\n",
    "            codonDict['AGT'] = 60\n",
    "        else:\n",
    "            codonDict[codon] = codonnum\n",
    "            codonnum += 1\n",
    "\n",
    "\n",
    "# to convert 3 letter codons to numbers between 0 and 63\n",
    "def get_numerical_codon_sequence(seq):\n",
    "    numseq = list()\n",
    "    for pos in range(0, len(seq) - 3, 3):\n",
    "        try:\n",
    "            numseq.append(str(codonDict[seq[pos:pos + 3]]))\n",
    "        except KeyError:\n",
    "            numseq.append('-1')\n",
    "            raise\n",
    "            return None\n",
    "    return ' '.join(numseq)\n",
    "\n",
    "\n",
    "# starting yfp sequence for leucine starvation expts\n",
    "yfp0 = Bio.SeqIO.read('../annotations/simulations/yfp0.fa', 'fasta')\n",
    "yfp0 = str(yfp0.seq)\n",
    "\n",
    "# starting sequence for serine starvation expts\n",
    "# all ser codons in yfp0 were AGC\n",
    "yfp_agc = list(yfp0)\n",
    "for pos in range(0, len(yfp0), 3):\n",
    "    current_codon = yfp0[pos:pos + 3]\n",
    "    if current_codon in SynonymousCodons['SER']:\n",
    "        yfp_agc[pos:pos + 3] = 'AGC'\n",
    "yfp_agc = ''.join(yfp_agc)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true
   },
   "outputs": [],
   "source": [
    "%matplotlib inline\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import pyfaidx\n",
    "import os\n",
    "import shutil\n",
    "import HTSeq\n",
    "import matplotlib.pyplot as plt\n",
    "import cPickle as pickle  # python native format storage library.\n",
    "import Bio.SeqIO  # dna sequence input and output.\n",
    "from Bio.SeqUtils import seq1  # for converting 3 letter amino acid code to 1 letter code\n",
    "from Bio.SeqUtils import CodonUsage  # provides dictionary of codon names\n",
    "from Bio.SeqUtils.CodonUsage import SynonymousCodons  # provides dictionary of codon names\n",
    "from Bio.SeqUtils import CodonUsageIndices  # provides the Ecoli Codon \n",
    "# Adaptation Index class.\n",
    "from Bio.SeqRecord import SeqRecord  # to read .gb and .fasta files.\n",
    "import re  # regular expression \n",
    "import numpy.random  # for creating random numbers\n",
    "from Bio import Entrez  # to retrieve records from NCBI\n",
    "import itertools  # iterator tools\n",
    "from IPython.core.display import clear_output, HTML  # for refreshing and displaying output\n",
    "import copy\n",
    "\n",
    "# create a dictionary of codon names and number (arranged alphabetically by aa)\n",
    "codonnum = 0\n",
    "codonDict = dict()\n",
    "for aa in sorted(SynonymousCodons, key=lambda aa3: seq1(aa3)):\n",
    "    if aa == 'STOP':\n",
    "        continue\n",
    "    for codon in sorted(SynonymousCodons[aa]):\n",
    "        # these two codons are numbered out of order for consistent notation\n",
    "        # with Subramaniam et al. Cell 2014\n",
    "        if codon in ['AGC']:\n",
    "            codonDict['AGC'] = 59\n",
    "        elif codon in ['AGT']:\n",
    "            codonDict['AGT'] = 60\n",
    "        else:\n",
    "            codonDict[codon] = codonnum\n",
    "            codonnum += 1\n",
    "\n",
    "\n",
    "# to convert 3 letter codons to numbers between 0 and 63\n",
    "def get_numerical_codon_sequence(seq):\n",
    "    numseq = list()\n",
    "    for pos in range(0, len(seq) - 3, 3):\n",
    "        try:\n",
    "            numseq.append(str(codonDict[seq[pos:pos + 3]]))\n",
    "        except KeyError:\n",
    "            numseq.append('-1')\n",
    "            raise\n",
    "            return None\n",
    "    return ' '.join(numseq)\n",
    "\n",
    "\n",
    "# starting yfp sequence for leucine starvation expts\n",
    "yfp0 = Bio.SeqIO.read('../annotations/simulations/yfp0.fa', 'fasta')\n",
    "yfp0 = str(yfp0.seq)\n",
    "\n",
    "# starting sequence for serine starvation expts\n",
    "# all ser codons in yfp0 were AGC\n",
    "yfp_agc = list(yfp0)\n",
    "for pos in range(0, len(yfp0), 3):\n",
    "    current_codon = yfp0[pos:pos + 3]\n",
    "    if current_codon in SynonymousCodons['SER']:\n",
    "        yfp_agc[pos:pos + 3] = 'AGC'\n",
    "yfp_agc = ''.join(yfp_agc)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": true,
    "editable": true
   },
   "source": [
    "### Create stall strength file for Run 3 initiation mutant prediction based on fit of Run 2 to single mutant data in 3 different models (for fig. 4A, fig. 4 supplement 1A-G)\n",
    "\n",
    "Also print elongation rate fit shown in Table S2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "tRNA accommodation rate at CTA67 (TJ model)\t0.07s-1\n",
      "tRNA accommodation rate at CTA140 (TJ model)\t0.12s-1\n",
      "tRNA accommodation rate at CTA200 (TJ model)\t0.107s-1\n",
      "tRNA accommodation rate at CTA45 (TJ model)\t0.091s-1\n",
      "tRNA accommodation rate at CTC67 (TJ model)\t0.184s-1\n",
      "tRNA accommodation rate at CTC140 (TJ model)\t0.211s-1\n",
      "tRNA accommodation rate at CTT67 (TJ model)\t0.0949s-1\n",
      "tRNA accommodation rate at CTT45 (TJ model)\t0.133s-1\n",
      "tRNA accommodation rate at CTA67 (CSAT model)\t0.0896s-1\n",
      "tRNA accommodation rate at CTA140 (CSAT model)\t0.173s-1\n",
      "tRNA accommodation rate at CTA200 (CSAT model)\t0.154s-1\n",
      "tRNA accommodation rate at CTA45 (CSAT model)\t0.129s-1\n",
      "tRNA accommodation rate at CTC67 (CSAT model)\t0.344s-1\n",
      "tRNA accommodation rate at CTC140 (CSAT model)\t0.444s-1\n",
      "tRNA accommodation rate at CTT67 (CSAT model)\t0.135s-1\n",
      "tRNA accommodation rate at CTT45 (CSAT model)\t0.205s-1\n",
      "tRNA accommodation rate at CTA67 (SAT model)\t0.292s-1\n",
      "tRNA accommodation rate at CTA140 (SAT model)\t0.637s-1\n",
      "tRNA accommodation rate at CTA200 (SAT model)\t0.55s-1\n",
      "tRNA accommodation rate at CTA45 (SAT model)\t0.444s-1\n",
      "tRNA accommodation rate at CTC67 (SAT model)\t1.45s-1\n",
      "tRNA accommodation rate at CTC140 (SAT model)\t1.83s-1\n",
      "tRNA accommodation rate at CTT67 (SAT model)\t0.469s-1\n",
      "tRNA accommodation rate at CTT45 (SAT model)\t0.713s-1\n"
     ]
    }
   ],
   "source": [
    "fitdatafolder = '../processeddata/simulations/'\n",
    "fitdatafiles = os.listdir(fitdatafolder)\n",
    "fitdatafiles = [\n",
    "    fitdatafolder + File for File in fitdatafiles\n",
    "    if File.startswith('run2_fit_stallstrength_for_initiation_')\n",
    "]\n",
    "\n",
    "# find the location of all leucine codons to convert leu codon serial number\n",
    "# to absolute position along yfp in codon units for simulation\n",
    "leupositions = dict()\n",
    "leucodon_number = 1\n",
    "for position in range(0, len(yfp0), 3):\n",
    "    currentcodon = yfp0[position:position + 3]\n",
    "    if currentcodon == 'CTG':\n",
    "        leupositions[leucodon_number] = position / 3\n",
    "        leucodon_number += 1\n",
    "\n",
    "codonnames = {23: 'CTA', 24: 'CTC', 26: 'CTT'}\n",
    "modelnames = {\n",
    "    'trafficjam': 'TJ',\n",
    "    'selpreterm': 'SAT',\n",
    "    '5primepreterm': 'CSAT',\n",
    "}\n",
    "\n",
    "for fitdatafile in fitdatafiles:\n",
    "    fitdata = pd.read_table(fitdatafile)\n",
    "    fitdata['codon'] = fitdata['mutant'].apply(lambda string: string[:3])\n",
    "    fitdata['pos'] = fitdata['mutant'].apply(lambda string: int(string[3:]))\n",
    "    fitdata = fitdata.drop(['mutant'], axis=1)\n",
    "    fitdata['pos'] = fitdata['pos'].apply(lambda pos: leupositions[pos])\n",
    "    fitdata['codon'] = fitdata['codon'].apply(\n",
    "        lambda codon: codonDict[codon.upper()])\n",
    "    fitdata = fitdata[['codon', 'pos', 'stallstrength']]\n",
    "    model = fitdatafile.split('_')[-1].split('.')[0]\n",
    "    fitdata.to_csv(\n",
    "        '../processeddata/simulations/run3_stallstrengthfits_' + model +\n",
    "        '.tsv',\n",
    "        sep='\\t',\n",
    "        index=False)\n",
    "    for row in fitdata.iterrows():\n",
    "        codon = row[1]['codon']\n",
    "        pos = row[1]['pos']\n",
    "        rate = row[1]['stallstrength']\n",
    "        print 'tRNA accommodation rate at {0}{1} ({2} model)\\t{3:.3}s-1'.format(\n",
    "            codonnames[codon], int(pos), modelnames[model], rate)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": true,
    "editable": true
   },
   "source": [
    "### Create mRNA sequence files for Run 3 to predict initiation mutant YFP synthesis rates in 3 different models  (for fig. 4A, fig. 4 supplement 1A-G)\n",
    "\n",
    "Also print initiation rate fit shown in Table S2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Inferred initiation rates (s-1)\n",
      "0.143\n",
      "0.219\n",
      "0.334\n",
      "0.101\n",
      "0.300\n"
     ]
    }
   ],
   "source": [
    "initiationRateFile = '../processeddata/platereader/inferred_initiationrates_for_initiation_simulations.tsv'\n",
    "\n",
    "mutation_locations = [\n",
    "    {\n",
    "        6: 'cta'\n",
    "    },\n",
    "    {\n",
    "        10: 'cta'\n",
    "    },\n",
    "    {\n",
    "        14: 'cta'\n",
    "    },\n",
    "    {\n",
    "        18: 'cta'\n",
    "    },\n",
    "    {\n",
    "        10: 'ctc'\n",
    "    },\n",
    "    {\n",
    "        14: 'ctc'\n",
    "    },\n",
    "    {\n",
    "        6: 'ctt'\n",
    "    },\n",
    "    {\n",
    "        10: 'ctt'\n",
    "    },\n",
    "]\n",
    "\n",
    "yfpmutants = dict()\n",
    "for mutant in mutation_locations:\n",
    "    key = '_'.join(['yfp'] + [\n",
    "        codon + str(location) for location, codon in mutant.items()\n",
    "    ])\n",
    "    yfpmutants[key] = list(yfp0)\n",
    "    leucodon_number = 0\n",
    "    for position in range(0, len(yfp0), 3):\n",
    "        currentcodon = yfp0[position:position + 3]\n",
    "        # proceed only if the codon is a Leu codon (which are all CTG in yfp0)\n",
    "        if currentcodon not in ['CTG']:\n",
    "            continue\n",
    "        leucodon_number += 1\n",
    "        for location in mutant.keys():\n",
    "            if leucodon_number == location:\n",
    "                yfpmutants[key][position:position + 3] = mutant[\n",
    "                    location].upper()\n",
    "    yfpmutants[key] = ''.join(yfpmutants[key])\n",
    "\n",
    "defaultMrnaCopyNumber = 1  # per cell\n",
    "defaultInitationRate = 0.3  # s-1, This is the median initiation rate\n",
    "\n",
    "experimentInitiationRates = pd.read_table(initiationRateFile, index_col=0)\n",
    "\n",
    "experimentInitiationRates[\n",
    "    'initiationRateForSimulation'] = experimentInitiationRates[\n",
    "        'inferredInitiationRate'].apply(lambda x: x * defaultInitationRate)\n",
    "\n",
    "print('Inferred initiation rates (s-1)')\n",
    "for initiationRate in experimentInitiationRates['initiationRateForSimulation']:\n",
    "    for mutant in yfpmutants:\n",
    "        outputFile = '../annotations/simulations/run3/' + \\\n",
    "        '%s_initiationrate_%0.4g.csv'%(mutant, initiationRate)\n",
    "        num_seq = ''.join(\n",
    "            get_numerical_codon_sequence(yfpmutants[mutant][:-3]))\n",
    "        File = open(outputFile, 'w')\n",
    "\n",
    "        File.write(\"%0.4g\\t%d\\t%s\\n\" %\n",
    "                   (initiationRate, defaultMrnaCopyNumber,\n",
    "                    get_numerical_codon_sequence(yfp0[:-3])))\n",
    "\n",
    "        File.write(\"%0.4g\\t%d\\t%s\\n\" % (initiationRate, defaultMrnaCopyNumber,\n",
    "                                        num_seq))\n",
    "    File.close()\n",
    "    print '%0.3f' % initiationRate"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Create stall strength file for Run 14 serine initiation mutant prediction based on fit of Run 13 to single mutant data in 3 different models  (for fig. 4 supplement 1H)\n",
    "\n",
    "Also print elongation rate fit shown in Table S2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "tRNA accommodation rate at TCG146 (TJ model)\t0.273s-1\n",
      "tRNA accommodation rate at TCG146 (CSAT model)\t0.767s-1\n",
      "tRNA accommodation rate at TCG146 (SAT model)\t4.59s-1\n"
     ]
    }
   ],
   "source": [
    "fitdatafolder = '../processeddata/simulations/'\n",
    "fitdatafiles = os.listdir(fitdatafolder)\n",
    "fitdatafiles = [\n",
    "    fitdatafolder + File for File in fitdatafiles\n",
    "    if File.startswith('run13_serine_fit_stallstrength_for_initiation_')\n",
    "]\n",
    "\n",
    "# find the location of all leucine codons to convert leu codon serial number\n",
    "# to absolute position along yfp in codon units for simulation\n",
    "serpositions = dict()\n",
    "sercodon_number = 1\n",
    "for position in range(0, len(yfp_agc), 3):\n",
    "    currentcodon = yfp_agc[position:position + 3]\n",
    "    if currentcodon == 'AGC':\n",
    "        serpositions[sercodon_number] = position / 3\n",
    "        sercodon_number += 1\n",
    "\n",
    "codonnames = {46: 'TCG'}\n",
    "modelnames = {\n",
    "    'trafficjam': 'TJ',\n",
    "    'selpreterm': 'SAT',\n",
    "    '5primepreterm': 'CSAT',\n",
    "}\n",
    "\n",
    "for fitdatafile in fitdatafiles:\n",
    "    fitdata = pd.read_table(fitdatafile)\n",
    "    fitdata['codon'] = fitdata['mutant'].apply(lambda string: string[:3])\n",
    "    fitdata['pos'] = fitdata['mutant'].apply(lambda string: int(string[3:]))\n",
    "    fitdata = fitdata.drop(['mutant'], axis=1)\n",
    "    fitdata['pos'] = fitdata['pos'].apply(lambda pos: serpositions[pos])\n",
    "    fitdata['codon'] = fitdata['codon'].apply(\n",
    "        lambda codon: codonDict[codon.upper()])\n",
    "    fitdata = fitdata[['codon', 'pos', 'stallstrength']]\n",
    "    model = fitdatafile.split('_')[-1].split('.')[0]\n",
    "    fitdata.to_csv(\n",
    "        '../processeddata/simulations/run14_stallstrengthfits_' + model +\n",
    "        '.tsv',\n",
    "        sep='\\t',\n",
    "        index=False)\n",
    "    for row in fitdata.iterrows():\n",
    "        codon = row[1]['codon']\n",
    "        pos = row[1]['pos']\n",
    "        rate = row[1]['stallstrength']\n",
    "        print 'tRNA accommodation rate at {0}{1} ({2} model)\\t{3:.3}s-1'.format(\n",
    "            codonnames[codon], int(pos), modelnames[model], rate)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Create mRNA sequence files for Run 14 to predict serine initiation mutant YFP synthesis rates in 3 different models (for fig. 4 supplement 1H)\n",
    "\n",
    "Also print initiation rate fit shown in Table S2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Inferred initiation rates (s-1)\n",
      "0.143\n",
      "0.219\n",
      "0.334\n",
      "0.101\n",
      "0.300\n"
     ]
    }
   ],
   "source": [
    "initiationRateFile = '../processeddata/platereader/inferred_initiationrates_for_initiation_simulations.tsv'\n",
    "\n",
    "mutation_locations = [\n",
    "    {\n",
    "        4: 'tcg'\n",
    "    },\n",
    "    {\n",
    "        5: 'tcg'\n",
    "    },\n",
    "    {\n",
    "        6: 'tcg'\n",
    "    },\n",
    "]\n",
    "\n",
    "yfpmutants = dict()\n",
    "for mutant in mutation_locations:\n",
    "    key = '_'.join(['yfp'] + [\n",
    "        codon + str(location) for location, codon in mutant.items()\n",
    "    ])\n",
    "    yfpmutants[key] = list(yfp_agc)\n",
    "    sercodon_number = 0\n",
    "    for position in range(0, len(yfp_agc), 3):\n",
    "        currentcodon = yfp_agc[position:position + 3]\n",
    "        # proceed only if the codon is a Ser codon (which are all AGC in yfp_agc)\n",
    "        if currentcodon not in ['AGC']:\n",
    "            continue\n",
    "        sercodon_number += 1\n",
    "        for location in mutant.keys():\n",
    "            if sercodon_number == location:\n",
    "                yfpmutants[key][position:position + 3] = mutant[\n",
    "                    location].upper()\n",
    "    yfpmutants[key] = ''.join(yfpmutants[key])\n",
    "\n",
    "defaultMrnaCopyNumber = 1  # per cell\n",
    "defaultInitationRate = 0.3  # s-1, This is the median initiation rate\n",
    "\n",
    "experimentInitiationRates = pd.read_table(initiationRateFile, index_col=0)\n",
    "\n",
    "experimentInitiationRates[\n",
    "    'initiationRateForSimulation'] = experimentInitiationRates[\n",
    "        'inferredInitiationRate'].apply(lambda x: x * defaultInitationRate)\n",
    "print('Inferred initiation rates (s-1)')\n",
    "for initiationRate in experimentInitiationRates['initiationRateForSimulation']:\n",
    "    for mutant in yfpmutants:\n",
    "        outputFile = '../annotations/simulations/run14/' + \\\n",
    "        '%s_initiationrate_%0.4g.csv'%(mutant, initiationRate)\n",
    "        num_seq = ''.join(\n",
    "            get_numerical_codon_sequence(yfpmutants[mutant][:-3]))\n",
    "        File = open(outputFile, 'w')\n",
    "\n",
    "        File.write(\"%0.4g\\t%d\\t%s\\n\" %\n",
    "                   (initiationRate, defaultMrnaCopyNumber,\n",
    "                    get_numerical_codon_sequence(yfp_agc[:-3])))\n",
    "\n",
    "        File.write(\"%0.4g\\t%d\\t%s\\n\" % (initiationRate, defaultMrnaCopyNumber,\n",
    "                                        num_seq))\n",
    "    File.close()\n",
    "    print '%0.3f' % initiationRate"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": true,
    "editable": true
   },
   "source": [
    "### Create stall strength file for Run 4 CTC, CTT double mutant prediction based on fit of Run 2 to single mutant data (for fig. 5 supplement 1A-B)\n",
    "\n",
    "Also print elongation rate fit shown in Table S3"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "tRNA accommodation rate at CTA67 (TJ model)\t0.046s-1\n",
      "tRNA accommodation rate at CTA140 (TJ model)\t0.0962s-1\n",
      "tRNA accommodation rate at CTA200 (TJ model)\t0.0914s-1\n",
      "tRNA accommodation rate at CTA14 (TJ model)\t0.0515s-1\n",
      "tRNA accommodation rate at CTA45 (TJ model)\t0.0667s-1\n",
      "tRNA accommodation rate at CTC67 (TJ model)\t0.267s-1\n",
      "tRNA accommodation rate at CTC140 (TJ model)\t0.286s-1\n",
      "tRNA accommodation rate at CTC200 (TJ model)\t0.284s-1\n",
      "tRNA accommodation rate at CTC14 (TJ model)\t0.137s-1\n",
      "tRNA accommodation rate at CTC45 (TJ model)\t0.205s-1\n",
      "tRNA accommodation rate at CTT67 (TJ model)\t0.139s-1\n",
      "tRNA accommodation rate at CTT140 (TJ model)\t0.27s-1\n",
      "tRNA accommodation rate at CTT200 (TJ model)\t0.246s-1\n",
      "tRNA accommodation rate at CTT14 (TJ model)\t0.125s-1\n",
      "tRNA accommodation rate at CTT45 (TJ model)\t0.172s-1\n",
      "tRNA accommodation rate at CTA67 (SAT model)\t0.186s-1\n",
      "tRNA accommodation rate at CTA140 (SAT model)\t0.464s-1\n",
      "tRNA accommodation rate at CTA200 (SAT model)\t0.425s-1\n",
      "tRNA accommodation rate at CTA14 (SAT model)\t0.231s-1\n",
      "tRNA accommodation rate at CTA45 (SAT model)\t0.287s-1\n",
      "tRNA accommodation rate at CTC67 (SAT model)\t3.06s-1\n",
      "tRNA accommodation rate at CTC140 (SAT model)\t4.4s-1\n",
      "tRNA accommodation rate at CTC200 (SAT model)\t4.6s-1\n",
      "tRNA accommodation rate at CTC14 (SAT model)\t0.764s-1\n",
      "tRNA accommodation rate at CTC45 (SAT model)\t1.59s-1\n",
      "tRNA accommodation rate at CTT67 (SAT model)\t0.777s-1\n",
      "tRNA accommodation rate at CTT140 (SAT model)\t3.94s-1\n",
      "tRNA accommodation rate at CTT200 (SAT model)\t2.85s-1\n",
      "tRNA accommodation rate at CTT14 (SAT model)\t0.649s-1\n",
      "tRNA accommodation rate at CTT45 (SAT model)\t1.09s-1\n",
      "tRNA accommodation rate at CTA67 (CSAT model)\t0.0585s-1\n",
      "tRNA accommodation rate at CTA140 (CSAT model)\t0.139s-1\n",
      "tRNA accommodation rate at CTA200 (CSAT model)\t0.13s-1\n",
      "tRNA accommodation rate at CTA14 (CSAT model)\t0.0649s-1\n",
      "tRNA accommodation rate at CTA45 (CSAT model)\t0.0875s-1\n",
      "tRNA accommodation rate at CTC67 (CSAT model)\t0.6s-1\n",
      "tRNA accommodation rate at CTC140 (CSAT model)\t0.817s-1\n",
      "tRNA accommodation rate at CTC200 (CSAT model)\t0.754s-1\n",
      "tRNA accommodation rate at CTC14 (CSAT model)\t0.193s-1\n",
      "tRNA accommodation rate at CTC45 (CSAT model)\t0.379s-1\n",
      "tRNA accommodation rate at CTT67 (CSAT model)\t0.213s-1\n",
      "tRNA accommodation rate at CTT140 (CSAT model)\t0.69s-1\n",
      "tRNA accommodation rate at CTT200 (CSAT model)\t0.582s-1\n",
      "tRNA accommodation rate at CTT14 (CSAT model)\t0.171s-1\n",
      "tRNA accommodation rate at CTT45 (CSAT model)\t0.285s-1\n"
     ]
    }
   ],
   "source": [
    "fitdatafolder = '../processeddata/simulations/'\n",
    "fitdatafiles = os.listdir(fitdatafolder)\n",
    "fitdatafiles = [\n",
    "    fitdatafolder + File for File in fitdatafiles\n",
    "    if File.startswith('run2_fit_stallstrength_for_double_')\n",
    "]\n",
    "\n",
    "# find the location of all leucine codons to convert leu codon serial number\n",
    "# to absolute position along yfp in codon units for simulation\n",
    "leupositions = dict()\n",
    "leucodon_number = 1\n",
    "for position in range(0, len(yfp0), 3):\n",
    "    currentcodon = yfp0[position:position + 3]\n",
    "    if currentcodon == 'CTG':\n",
    "        leupositions[leucodon_number] = position / 3\n",
    "        leucodon_number += 1\n",
    "\n",
    "codonnames = {23: 'CTA', 24: 'CTC', 26: 'CTT'}\n",
    "modelnames = {\n",
    "    'trafficjam': 'TJ',\n",
    "    'selpreterm': 'SAT',\n",
    "    '5primepreterm': 'CSAT',\n",
    "}\n",
    "\n",
    "for fitdatafile in fitdatafiles:\n",
    "    fitdata = pd.read_table(fitdatafile)\n",
    "    fitdata['codon'] = fitdata['mutant'].apply(lambda string: string[:3])\n",
    "    fitdata['pos'] = fitdata['mutant'].apply(lambda string: int(string[3:]))\n",
    "    fitdata = fitdata.drop(['mutant'], axis=1)\n",
    "    fitdata['pos'] = fitdata['pos'].apply(lambda pos: leupositions[pos])\n",
    "    fitdata['codon'] = fitdata['codon'].apply(\n",
    "        lambda codon: codonDict[codon.upper()])\n",
    "    fitdata = fitdata[['codon', 'pos', 'stallstrength']]\n",
    "    model = fitdatafile.split('_')[-1].split('.')[0]\n",
    "    fitdata.to_csv(\n",
    "        '../processeddata/simulations/run4_stallstrengthfits_' + model +\n",
    "        '.tsv',\n",
    "        sep='\\t',\n",
    "        index=False)\n",
    "    for row in fitdata.iterrows():\n",
    "        codon = row[1]['codon']\n",
    "        pos = row[1]['pos']\n",
    "        rate = row[1]['stallstrength']\n",
    "        print 'tRNA accommodation rate at {0}{1} ({2} model)\\t{3:.3}s-1'.format(\n",
    "            codonnames[codon], int(pos), modelnames[model], rate)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": true,
    "editable": true
   },
   "source": [
    "### Create mRNA sequence files for Run 4 to predict CTC CTT double mutant YFP synthesis rates in different models (for fig. 5 supplement 1A-B)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true
   },
   "outputs": [],
   "source": [
    "mutation_locations = [\n",
    "    {\n",
    "        2: 'ctc'\n",
    "    },\n",
    "    {\n",
    "        6: 'ctc'\n",
    "    },\n",
    "    {\n",
    "        10: 'ctc'\n",
    "    },\n",
    "    {\n",
    "        14: 'ctc'\n",
    "    },\n",
    "    {\n",
    "        18: 'ctc'\n",
    "    },\n",
    "    {\n",
    "        2: 'ctc',\n",
    "        6: 'ctc'\n",
    "    },\n",
    "    {\n",
    "        2: 'ctc',\n",
    "        10: 'ctc'\n",
    "    },\n",
    "    {\n",
    "        2: 'ctc',\n",
    "        14: 'ctc'\n",
    "    },\n",
    "    {\n",
    "        2: 'ctc',\n",
    "        18: 'ctc'\n",
    "    },\n",
    "    {\n",
    "        6: 'ctc',\n",
    "        10: 'ctc'\n",
    "    },\n",
    "    {\n",
    "        6: 'ctc',\n",
    "        14: 'ctc'\n",
    "    },\n",
    "    {\n",
    "        6: 'ctc',\n",
    "        18: 'ctc'\n",
    "    },\n",
    "    {\n",
    "        10: 'ctc',\n",
    "        14: 'ctc'\n",
    "    },\n",
    "    {\n",
    "        10: 'ctc',\n",
    "        18: 'ctc'\n",
    "    },\n",
    "    {\n",
    "        14: 'ctc',\n",
    "        18: 'ctc'\n",
    "    },\n",
    "    {\n",
    "        2: 'ctt'\n",
    "    },\n",
    "    {\n",
    "        6: 'ctt'\n",
    "    },\n",
    "    {\n",
    "        10: 'ctt'\n",
    "    },\n",
    "    {\n",
    "        14: 'ctt'\n",
    "    },\n",
    "    {\n",
    "        18: 'ctt'\n",
    "    },\n",
    "    {\n",
    "        2: 'ctt',\n",
    "        6: 'ctt'\n",
    "    },\n",
    "    {\n",
    "        2: 'ctt',\n",
    "        10: 'ctt'\n",
    "    },\n",
    "    {\n",
    "        2: 'ctt',\n",
    "        14: 'ctt'\n",
    "    },\n",
    "    {\n",
    "        2: 'ctt',\n",
    "        18: 'ctt'\n",
    "    },\n",
    "    {\n",
    "        6: 'ctt',\n",
    "        10: 'ctt'\n",
    "    },\n",
    "    {\n",
    "        6: 'ctt',\n",
    "        14: 'ctt'\n",
    "    },\n",
    "    {\n",
    "        6: 'ctt',\n",
    "        18: 'ctt'\n",
    "    },\n",
    "    {\n",
    "        10: 'ctt',\n",
    "        14: 'ctt'\n",
    "    },\n",
    "    {\n",
    "        10: 'ctt',\n",
    "        18: 'ctt'\n",
    "    },\n",
    "    {\n",
    "        14: 'ctt',\n",
    "        18: 'ctt'\n",
    "    },\n",
    "]\n",
    "\n",
    "yfpmutants = dict()\n",
    "for mutant in mutation_locations:\n",
    "    key = '_'.join(['yfp'] + [\n",
    "        codon + str(location) for location, codon in mutant.items()\n",
    "    ])\n",
    "    yfpmutants[key] = list(yfp0)\n",
    "    leucodon_number = 0\n",
    "    for position in range(0, len(yfp0), 3):\n",
    "        currentcodon = yfp0[position:position + 3]\n",
    "        # proceed only if the codon is a Leu codon (which are all CTG in yfp0)\n",
    "        if currentcodon not in ['CTG']:\n",
    "            continue\n",
    "        leucodon_number += 1\n",
    "        for location in mutant.keys():\n",
    "            if leucodon_number == location:\n",
    "                yfpmutants[key][position:position + 3] = mutant[\n",
    "                    location].upper()\n",
    "    yfpmutants[key] = ''.join(yfpmutants[key])\n",
    "\n",
    "defaultMrnaCopyNumber = 1  # per cell\n",
    "defaultInitationRate = 0.3  # s-1, This is the median initiation rate\n",
    "\n",
    "listOfInitiationRates = [defaultInitationRate]\n",
    "\n",
    "for initiationRate in listOfInitiationRates:\n",
    "    for mutant in yfpmutants:\n",
    "        outputFile = '../annotations/simulations/run4/' + \\\n",
    "        '%s_initiationrate_%0.4g.csv'%(mutant, initiationRate)\n",
    "        num_seq = ''.join(\n",
    "            get_numerical_codon_sequence(yfpmutants[mutant][:-3]))\n",
    "        File = open(outputFile, 'w')\n",
    "\n",
    "        File.write(\"%0.4g\\t%d\\t%s\\n\" %\n",
    "                   (initiationRate, defaultMrnaCopyNumber,\n",
    "                    get_numerical_codon_sequence(yfp0[:-3])))\n",
    "\n",
    "        File.write(\"%0.4g\\t%d\\t%s\\n\" % (initiationRate, defaultMrnaCopyNumber,\n",
    "                                        num_seq))\n",
    "    File.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Create stall strength file for Run 16 CTA multiple mutant prediction based on fit of Run 2 to single mutant data (for fig. 5)\n",
    "\n",
    "Also print elongation rate fit shown in Table S3"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "tRNA accommodation rate at CTA67 (SAT model)\t0.26s-1\n",
      "tRNA accommodation rate at CTA140 (SAT model)\t0.513s-1\n",
      "tRNA accommodation rate at CTA200 (SAT model)\t0.404s-1\n",
      "tRNA accommodation rate at CTA14 (SAT model)\t0.266s-1\n",
      "tRNA accommodation rate at CTA45 (SAT model)\t0.376s-1\n",
      "tRNA accommodation rate at CTA67 (TJ model)\t0.0617s-1\n",
      "tRNA accommodation rate at CTA140 (TJ model)\t0.1s-1\n",
      "tRNA accommodation rate at CTA200 (TJ model)\t0.0876s-1\n",
      "tRNA accommodation rate at CTA14 (TJ model)\t0.0567s-1\n",
      "tRNA accommodation rate at CTA45 (TJ model)\t0.0837s-1\n",
      "tRNA accommodation rate at CTA67 (CSAT model)\t0.0769s-1\n",
      "tRNA accommodation rate at CTA140 (CSAT model)\t0.147s-1\n",
      "tRNA accommodation rate at CTA200 (CSAT model)\t0.123s-1\n",
      "tRNA accommodation rate at CTA14 (CSAT model)\t0.0726s-1\n",
      "tRNA accommodation rate at CTA45 (CSAT model)\t0.113s-1\n"
     ]
    }
   ],
   "source": [
    "fitdatafolder = '../processeddata/simulations/'\n",
    "fitdatafiles = os.listdir(fitdatafolder)\n",
    "fitdatafiles = [\n",
    "    fitdatafolder + File for File in fitdatafiles\n",
    "    if File.startswith('run2_fit_stallstrength_for_leucine_multiple_')\n",
    "]\n",
    "\n",
    "# find the location of all leucine codons to convert leu codon serial number\n",
    "# to absolute position along yfp in codon units for simulation\n",
    "leupositions = dict()\n",
    "leucodon_number = 1\n",
    "for position in range(0, len(yfp0), 3):\n",
    "    currentcodon = yfp0[position:position + 3]\n",
    "    if currentcodon == 'CTG':\n",
    "        leupositions[leucodon_number] = position / 3\n",
    "        leucodon_number += 1\n",
    "\n",
    "codonnames = {23: 'CTA', 24: 'CTC', 26: 'CTT'}\n",
    "modelnames = {\n",
    "    'trafficjam': 'TJ',\n",
    "    'selpreterm': 'SAT',\n",
    "    '5primepreterm': 'CSAT',\n",
    "}\n",
    "\n",
    "for fitdatafile in fitdatafiles:\n",
    "    fitdata = pd.read_table(fitdatafile)\n",
    "    fitdata['codon'] = fitdata['mutant'].apply(lambda string: string[:3])\n",
    "    fitdata['pos'] = fitdata['mutant'].apply(lambda string: int(string[3:]))\n",
    "    fitdata = fitdata.drop(['mutant'], axis=1)\n",
    "    fitdata['pos'] = fitdata['pos'].apply(lambda pos: leupositions[pos])\n",
    "    fitdata['codon'] = fitdata['codon'].apply(\n",
    "        lambda codon: codonDict[codon.upper()])\n",
    "    fitdata = fitdata[['codon', 'pos', 'stallstrength']]\n",
    "    model = fitdatafile.split('_')[-1].split('.')[0]\n",
    "    fitdata.to_csv(\n",
    "        '../processeddata/simulations/run16_stallstrengthfits_' + model +\n",
    "        '.tsv',\n",
    "        sep='\\t',\n",
    "        index=False)\n",
    "    for row in fitdata.iterrows():\n",
    "        codon = row[1]['codon']\n",
    "        pos = row[1]['pos']\n",
    "        rate = row[1]['stallstrength']\n",
    "        print 'tRNA accommodation rate at {0}{1} ({2} model)\\t{3:.3}s-1'.format(\n",
    "            codonnames[codon], int(pos), modelnames[model], rate)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Create mRNA sequence files for Run 16 to predict multiple mutant YFP synthesis rates in different models (for fig. 5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "mutation_locations = [\n",
    "    {\n",
    "        2: 'cta'\n",
    "    },\n",
    "    {\n",
    "        6: 'cta'\n",
    "    },\n",
    "    {\n",
    "        10: 'cta'\n",
    "    },\n",
    "    {\n",
    "        14: 'cta'\n",
    "    },\n",
    "    {\n",
    "        18: 'cta'\n",
    "    },\n",
    "    {\n",
    "        2: 'cta',\n",
    "        6: 'cta'\n",
    "    },\n",
    "    {\n",
    "        2: 'cta',\n",
    "        10: 'cta'\n",
    "    },\n",
    "    {\n",
    "        2: 'cta',\n",
    "        14: 'cta'\n",
    "    },\n",
    "    {\n",
    "        2: 'cta',\n",
    "        18: 'cta'\n",
    "    },\n",
    "    {\n",
    "        6: 'cta',\n",
    "        10: 'cta'\n",
    "    },\n",
    "    {\n",
    "        6: 'cta',\n",
    "        14: 'cta'\n",
    "    },\n",
    "    {\n",
    "        6: 'cta',\n",
    "        18: 'cta'\n",
    "    },\n",
    "    {\n",
    "        10: 'cta',\n",
    "        14: 'cta'\n",
    "    },\n",
    "    {\n",
    "        10: 'cta',\n",
    "        18: 'cta'\n",
    "    },\n",
    "    {\n",
    "        14: 'cta',\n",
    "        18: 'cta'\n",
    "    },\n",
    "    {\n",
    "        6: 'cta',\n",
    "        10: 'cta',\n",
    "        14: 'cta'\n",
    "    },\n",
    "    {\n",
    "        6: 'cta',\n",
    "        10: 'cta',\n",
    "        18: 'cta'\n",
    "    },\n",
    "    {\n",
    "        6: 'cta',\n",
    "        10: 'cta',\n",
    "        14: 'cta',\n",
    "        18: 'cta'\n",
    "    },\n",
    "]\n",
    "\n",
    "yfpmutants = dict()\n",
    "for mutant in mutation_locations:\n",
    "    key = '_'.join(['yfp'] + [\n",
    "        codon + str(location) for location, codon in mutant.items()\n",
    "    ])\n",
    "    yfpmutants[key] = list(yfp0)\n",
    "    leucodon_number = 0\n",
    "    for position in range(0, len(yfp0), 3):\n",
    "        currentcodon = yfp0[position:position + 3]\n",
    "        # proceed only if the codon is a Leu codon (which are all CTG in yfp0)\n",
    "        if currentcodon not in ['CTG']:\n",
    "            continue\n",
    "        leucodon_number += 1\n",
    "        for location in mutant.keys():\n",
    "            if leucodon_number == location:\n",
    "                yfpmutants[key][position:position + 3] = mutant[\n",
    "                    location].upper()\n",
    "    yfpmutants[key] = ''.join(yfpmutants[key])\n",
    "\n",
    "defaultMrnaCopyNumber = 1  # per cell\n",
    "defaultInitationRate = 0.3  # s-1, This is the median initiation rate\n",
    "\n",
    "listOfInitiationRates = [defaultInitationRate]\n",
    "\n",
    "for initiationRate in listOfInitiationRates:\n",
    "    for mutant in yfpmutants:\n",
    "        outputFile = '../annotations/simulations/run16/' + \\\n",
    "        '%s_initiationrate_%0.4g.csv'%(mutant, initiationRate)\n",
    "        num_seq = ''.join(\n",
    "            get_numerical_codon_sequence(yfpmutants[mutant][:-3]))\n",
    "        File = open(outputFile, 'w')\n",
    "\n",
    "        File.write(\"%0.4g\\t%d\\t%s\\n\" %\n",
    "                   (initiationRate, defaultMrnaCopyNumber,\n",
    "                    get_numerical_codon_sequence(yfp0[:-3])))\n",
    "\n",
    "        File.write(\"%0.4g\\t%d\\t%s\\n\" % (initiationRate, defaultMrnaCopyNumber,\n",
    "                                        num_seq))\n",
    "    File.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Create stall strength file for Run 15 serine double mutant prediction based on fit of Run 13 to single mutant data (for fig. 5 supplement 1C)\n",
    "\n",
    "Also print elongation rate fit shown in Table S3"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "tRNA accommodation rate at TCG27 (TJ model)\t0.263s-1\n",
      "tRNA accommodation rate at TCG29 (TJ model)\t0.266s-1\n",
      "tRNA accommodation rate at TCG85 (TJ model)\t0.281s-1\n",
      "tRNA accommodation rate at TCG146 (TJ model)\t0.273s-1\n",
      "tRNA accommodation rate at TCG201 (TJ model)\t0.252s-1\n",
      "tRNA accommodation rate at TCG204 (TJ model)\t0.29s-1\n",
      "tRNA accommodation rate at TCG27 (CSAT model)\t0.488s-1\n",
      "tRNA accommodation rate at TCG29 (CSAT model)\t0.509s-1\n",
      "tRNA accommodation rate at TCG85 (CSAT model)\t0.725s-1\n",
      "tRNA accommodation rate at TCG146 (CSAT model)\t0.767s-1\n",
      "tRNA accommodation rate at TCG201 (CSAT model)\t0.601s-1\n",
      "tRNA accommodation rate at TCG204 (CSAT model)\t0.935s-1\n",
      "tRNA accommodation rate at TCG27 (SAT model)\t2.19s-1\n",
      "tRNA accommodation rate at TCG29 (SAT model)\t2.13s-1\n",
      "tRNA accommodation rate at TCG85 (SAT model)\t3.82s-1\n",
      "tRNA accommodation rate at TCG146 (SAT model)\t4.59s-1\n",
      "tRNA accommodation rate at TCG201 (SAT model)\t3.22s-1\n",
      "tRNA accommodation rate at TCG204 (SAT model)\t5.15s-1\n"
     ]
    }
   ],
   "source": [
    "fitdatafolder = '../processeddata/simulations/'\n",
    "fitdatafiles = os.listdir(fitdatafolder)\n",
    "fitdatafiles = [\n",
    "    fitdatafolder + File for File in fitdatafiles\n",
    "    if File.startswith('run13_serine_fit_stallstrength_for_double_')\n",
    "]\n",
    "\n",
    "# find the location of all serine codons to convert ser codon serial number\n",
    "# to absolute position along yfp in codon units for simulation\n",
    "serpositions = dict()\n",
    "sercodon_number = 1\n",
    "for position in range(0, len(yfp_agc), 3):\n",
    "    currentcodon = yfp_agc[position:position + 3]\n",
    "    if currentcodon == 'AGC':\n",
    "        serpositions[sercodon_number] = position / 3\n",
    "        sercodon_number += 1\n",
    "\n",
    "codonnames = {46: 'TCG'}\n",
    "modelnames = {\n",
    "    'trafficjam': 'TJ',\n",
    "    'selpreterm': 'SAT',\n",
    "    '5primepreterm': 'CSAT',\n",
    "}\n",
    "\n",
    "for fitdatafile in fitdatafiles:\n",
    "    fitdata = pd.read_table(fitdatafile)\n",
    "    fitdata['codon'] = fitdata['mutant'].apply(lambda string: string[:3])\n",
    "    fitdata['pos'] = fitdata['mutant'].apply(lambda string: int(string[3:]))\n",
    "    fitdata = fitdata.drop(['mutant'], axis=1)\n",
    "    fitdata['pos'] = fitdata['pos'].apply(lambda pos: serpositions[pos])\n",
    "    fitdata['codon'] = fitdata['codon'].apply(\n",
    "        lambda codon: codonDict[codon.upper()])\n",
    "    fitdata = fitdata[['codon', 'pos', 'stallstrength']]\n",
    "    model = fitdatafile.split('_')[-1].split('.')[0]\n",
    "    fitdata.to_csv(\n",
    "        '../processeddata/simulations/run15_stallstrengthfits_' + model +\n",
    "        '.tsv',\n",
    "        sep='\\t',\n",
    "        index=False)\n",
    "    for row in fitdata.iterrows():\n",
    "        codon = row[1]['codon']\n",
    "        pos = row[1]['pos']\n",
    "        rate = row[1]['stallstrength']\n",
    "        print 'tRNA accommodation rate at {0}{1} ({2} model)\\t{3:.3}s-1'.format(\n",
    "            codonnames[codon], int(pos), modelnames[model], rate)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Create mRNA sequence files for Run 15 to predict  serine double mutant YFP synthesis rates in different models (for fig. 5 supplement 1C)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "mutation_locations = [\n",
    "    {\n",
    "        2: 'tcg'\n",
    "    },\n",
    "    {\n",
    "        3: 'tcg'\n",
    "    },\n",
    "    {\n",
    "        4: 'tcg'\n",
    "    },\n",
    "    {\n",
    "        5: 'tcg'\n",
    "    },\n",
    "    {\n",
    "        6: 'tcg'\n",
    "    },\n",
    "    {\n",
    "        7: 'tcg'\n",
    "    },\n",
    "    {\n",
    "        2: 'tcg',\n",
    "        5: 'tcg'\n",
    "    },\n",
    "    {\n",
    "        2: 'tcg',\n",
    "        6: 'tcg'\n",
    "    },\n",
    "    {\n",
    "        2: 'tcg',\n",
    "        7: 'tcg'\n",
    "    },\n",
    "    {\n",
    "        3: 'tcg',\n",
    "        5: 'tcg'\n",
    "    },\n",
    "    {\n",
    "        3: 'tcg',\n",
    "        6: 'tcg'\n",
    "    },\n",
    "    {\n",
    "        3: 'tcg',\n",
    "        7: 'tcg'\n",
    "    },\n",
    "    {\n",
    "        4: 'tcg',\n",
    "        5: 'tcg'\n",
    "    },\n",
    "    {\n",
    "        4: 'tcg',\n",
    "        6: 'tcg'\n",
    "    },\n",
    "    {\n",
    "        4: 'tcg',\n",
    "        7: 'tcg'\n",
    "    },\n",
    "]\n",
    "\n",
    "yfpmutants = dict()\n",
    "for mutant in mutation_locations:\n",
    "    key = '_'.join(['yfp'] + [\n",
    "        codon + str(location) for location, codon in mutant.items()\n",
    "    ])\n",
    "    yfpmutants[key] = list(yfp_agc)\n",
    "    sercodon_number = 0\n",
    "    for position in range(0, len(yfp_agc), 3):\n",
    "        currentcodon = yfp_agc[position:position + 3]\n",
    "        # proceed only if the codon is a Ser codon (which are all AGC in yfp_agc)\n",
    "        if currentcodon not in ['AGC']:\n",
    "            continue\n",
    "        sercodon_number += 1\n",
    "        for location in mutant.keys():\n",
    "            if sercodon_number == location:\n",
    "                yfpmutants[key][position:position + 3] = mutant[\n",
    "                    location].upper()\n",
    "    yfpmutants[key] = ''.join(yfpmutants[key])\n",
    "\n",
    "defaultMrnaCopyNumber = 1  # per cell\n",
    "defaultInitationRate = 0.3  # s-1, This is the median initiation rate\n",
    "\n",
    "listOfInitiationRates = [defaultInitationRate]\n",
    "\n",
    "for initiationRate in listOfInitiationRates:\n",
    "    for mutant in yfpmutants:\n",
    "        outputFile = '../annotations/simulations/run15/' + \\\n",
    "        '%s_initiationrate_%0.4g.csv'%(mutant, initiationRate)\n",
    "        num_seq = ''.join(\n",
    "            get_numerical_codon_sequence(yfpmutants[mutant][:-3]))\n",
    "        File = open(outputFile, 'w')\n",
    "\n",
    "        File.write(\"%0.4g\\t%d\\t%s\\n\" %\n",
    "                   (initiationRate, defaultMrnaCopyNumber,\n",
    "                    get_numerical_codon_sequence(yfp_agc[:-3])))\n",
    "\n",
    "        File.write(\"%0.4g\\t%d\\t%s\\n\" % (initiationRate, defaultMrnaCopyNumber,\n",
    "                                        num_seq))\n",
    "    File.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": true,
    "editable": true
   },
   "source": [
    "### Create stall strength file for Run 5 CTC distance mutant prediction based on fit of Run 2 to single mutant data (for fig. 6 supplement 1)\n",
    "\n",
    "Also print elongation rate fit shown in Table S5"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "tRNA accommodation rate at CTC67 (TJ model)\t0.226s-1\n",
      "tRNA accommodation rate at CTC118 (TJ model)\t0.199s-1\n",
      "tRNA accommodation rate at CTC124 (TJ model)\t0.205s-1\n",
      "tRNA accommodation rate at CTC136 (TJ model)\t0.236s-1\n",
      "tRNA accommodation rate at CTC140 (TJ model)\t0.273s-1\n",
      "tRNA accommodation rate at CTC59 (TJ model)\t0.228s-1\n",
      "tRNA accommodation rate at CTC63 (TJ model)\t0.214s-1\n",
      "tRNA accommodation rate at CTC67 (CSAT model)\t0.468s-1\n",
      "tRNA accommodation rate at CTC118 (CSAT model)\t0.387s-1\n",
      "tRNA accommodation rate at CTC124 (CSAT model)\t0.394s-1\n",
      "tRNA accommodation rate at CTC136 (CSAT model)\t0.512s-1\n",
      "tRNA accommodation rate at CTC140 (CSAT model)\t0.777s-1\n",
      "tRNA accommodation rate at CTC59 (CSAT model)\t0.483s-1\n",
      "tRNA accommodation rate at CTC63 (CSAT model)\t0.433s-1\n",
      "tRNA accommodation rate at CTC67 (SAT model)\t2.17s-1\n",
      "tRNA accommodation rate at CTC118 (SAT model)\t1.71s-1\n",
      "tRNA accommodation rate at CTC124 (SAT model)\t1.67s-1\n",
      "tRNA accommodation rate at CTC136 (SAT model)\t2.51s-1\n",
      "tRNA accommodation rate at CTC140 (SAT model)\t4.26s-1\n",
      "tRNA accommodation rate at CTC59 (SAT model)\t2.15s-1\n",
      "tRNA accommodation rate at CTC63 (SAT model)\t1.85s-1\n"
     ]
    }
   ],
   "source": [
    "fitdatafolder = '../processeddata/simulations/'\n",
    "fitdatafiles = os.listdir(fitdatafolder)\n",
    "fitdatafiles = [\n",
    "    fitdatafolder + File for File in fitdatafiles\n",
    "    if File.startswith('run2_fit_stallstrength_for_ctc_distance_')\n",
    "]\n",
    "\n",
    "# find the location of all leucine codons to convert leu codon serial number\n",
    "# to absolute position along yfp in codon units for simulation\n",
    "leupositions = dict()\n",
    "leucodon_number = 1\n",
    "for position in range(0, len(yfp0), 3):\n",
    "    currentcodon = yfp0[position:position + 3]\n",
    "    if currentcodon == 'CTG':\n",
    "        leupositions[leucodon_number] = position / 3\n",
    "        leucodon_number += 1\n",
    "\n",
    "codonnames = {23: 'CTA', 24: 'CTC', 26: 'CTT'}\n",
    "modelnames = {\n",
    "    'trafficjam': 'TJ',\n",
    "    'selpreterm': 'SAT',\n",
    "    '5primepreterm': 'CSAT',\n",
    "}\n",
    "\n",
    "for fitdatafile in fitdatafiles:\n",
    "    fitdata = pd.read_table(fitdatafile)\n",
    "    fitdata['codon'] = fitdata['mutant'].apply(lambda string: string[:3])\n",
    "    fitdata['pos'] = fitdata['mutant'].apply(lambda string: int(string[3:]))\n",
    "    fitdata = fitdata.drop(['mutant'], axis=1)\n",
    "    fitdata['pos'] = fitdata['pos'].apply(lambda pos: leupositions[pos])\n",
    "    fitdata['codon'] = fitdata['codon'].apply(\n",
    "        lambda codon: codonDict[codon.upper()])\n",
    "    fitdata = fitdata[['codon', 'pos', 'stallstrength']]\n",
    "    model = fitdatafile.split('_')[-1].split('.')[0]\n",
    "    fitdata.to_csv(\n",
    "        '../processeddata/simulations/run5_stallstrengthfits_' + model +\n",
    "        '.tsv',\n",
    "        sep='\\t',\n",
    "        index=False)\n",
    "    for row in fitdata.iterrows():\n",
    "        codon = row[1]['codon']\n",
    "        pos = row[1]['pos']\n",
    "        rate = row[1]['stallstrength']\n",
    "        print 'tRNA accommodation rate at {0}{1} ({2} model)\\t{3:.3}s-1'.format(\n",
    "            codonnames[codon], int(pos), modelnames[model], rate)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": true,
    "editable": true
   },
   "source": [
    "### Create mRNA sequence files for Run 5 to predict CTC distance mutant YFP synthesis rates in different models (for fig. 6 supplement 1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {
    "collapsed": true,
    "deletable": true,
    "editable": true
   },
   "outputs": [],
   "source": [
    "mutation_locations = [\n",
    "    {\n",
    "        8: 'ctc'\n",
    "    },\n",
    "    {\n",
    "        9: 'ctc'\n",
    "    },\n",
    "    {\n",
    "        10: 'ctc'\n",
    "    },\n",
    "    {\n",
    "        11: 'ctc'\n",
    "    },\n",
    "    {\n",
    "        12: 'ctc'\n",
    "    },\n",
    "    {\n",
    "        13: 'ctc'\n",
    "    },\n",
    "    {\n",
    "        14: 'ctc'\n",
    "    },\n",
    "    {\n",
    "        8: 'ctc',\n",
    "        9: 'ctc'\n",
    "    },\n",
    "    {\n",
    "        8: 'ctc',\n",
    "        10: 'ctc'\n",
    "    },\n",
    "    {\n",
    "        8: 'ctc',\n",
    "        11: 'ctc'\n",
    "    },\n",
    "    {\n",
    "        8: 'ctc',\n",
    "        12: 'ctc'\n",
    "    },\n",
    "    {\n",
    "        8: 'ctc',\n",
    "        13: 'ctc'\n",
    "    },\n",
    "    {\n",
    "        8: 'ctc',\n",
    "        14: 'ctc'\n",
    "    },\n",
    "    {\n",
    "        9: 'ctc',\n",
    "        10: 'ctc'\n",
    "    },\n",
    "    {\n",
    "        9: 'ctc',\n",
    "        11: 'ctc'\n",
    "    },\n",
    "    {\n",
    "        9: 'ctc',\n",
    "        12: 'ctc'\n",
    "    },\n",
    "    {\n",
    "        9: 'ctc',\n",
    "        13: 'ctc'\n",
    "    },\n",
    "    {\n",
    "        9: 'ctc',\n",
    "        14: 'ctc'\n",
    "    },\n",
    "    {\n",
    "        11: 'ctc',\n",
    "        12: 'ctc'\n",
    "    },\n",
    "    {\n",
    "        13: 'ctc',\n",
    "        14: 'ctc'\n",
    "    },\n",
    "]\n",
    "\n",
    "yfpmutants = dict()\n",
    "for mutant in mutation_locations:\n",
    "    key = '_'.join(['yfp'] + [\n",
    "        codon + str(location) for location, codon in mutant.items()\n",
    "    ])\n",
    "    yfpmutants[key] = list(yfp0)\n",
    "    leucodon_number = 0\n",
    "    for position in range(0, len(yfp0), 3):\n",
    "        currentcodon = yfp0[position:position + 3]\n",
    "        # proceed only if the codon is a Leu codon (which are all CTG in yfp0)\n",
    "        if currentcodon not in ['CTG']:\n",
    "            continue\n",
    "        leucodon_number += 1\n",
    "        for location in mutant.keys():\n",
    "            if leucodon_number == location:\n",
    "                yfpmutants[key][position:position + 3] = mutant[\n",
    "                    location].upper()\n",
    "    yfpmutants[key] = ''.join(yfpmutants[key])\n",
    "\n",
    "defaultMrnaCopyNumber = 1  # per cell\n",
    "defaultInitationRate = 0.3  # s-1, This is the median initiation rate\n",
    "\n",
    "listOfInitiationRates = [defaultInitationRate]\n",
    "\n",
    "for initiationRate in listOfInitiationRates:\n",
    "    for mutant in yfpmutants:\n",
    "        outputFile = '../annotations/simulations/run5/' + \\\n",
    "        '%s_initiationrate_%0.4g.csv'%(mutant, initiationRate)\n",
    "        num_seq = ''.join(\n",
    "            get_numerical_codon_sequence(yfpmutants[mutant][:-3]))\n",
    "        File = open(outputFile, 'w')\n",
    "\n",
    "        File.write(\"%0.4g\\t%d\\t%s\\n\" %\n",
    "                   (initiationRate, defaultMrnaCopyNumber,\n",
    "                    get_numerical_codon_sequence(yfp0[:-3])))\n",
    "\n",
    "        File.write(\"%0.4g\\t%d\\t%s\\n\" % (initiationRate, defaultMrnaCopyNumber,\n",
    "                                        num_seq))\n",
    "    File.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": true,
    "editable": true
   },
   "source": [
    "### Create stall strength file for Runs 6,7,8 to systematically vary initiation rate, number of stall sites and distance between stall sites in different models (for Fig. 3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true
   },
   "outputs": [],
   "source": [
    "fitdatafolder = '../processeddata/simulations/'\n",
    "fitdatafiles = os.listdir(fitdatafolder)\n",
    "fitdatafiles = [\n",
    "    fitdatafolder + File for File in fitdatafiles\n",
    "    if File.startswith('run4_stallstrengthfits_')\n",
    "]\n",
    "\n",
    "for fitdatafile in fitdatafiles:\n",
    "    fitdata = pd.read_table(fitdatafile)\n",
    "    fitdata = fitdata[['codon', 'pos', 'stallstrength']]\n",
    "    fitdata = fitdata[fitdata['codon'] == 23]  # CTA codon\n",
    "    # Use CTA 6 (Leu45) values for all codons\n",
    "    commonstallstrength = fitdata[fitdata['pos'] == 45].reset_index().ix[0][\n",
    "        'stallstrength']\n",
    "    newdataframe = pd.DataFrame(\n",
    "        [[23, loop, commonstallstrength] for loop in range(1, 238)],\n",
    "        columns=fitdata.columns)\n",
    "    model = fitdatafile.split('_')[-1].split('.')[0]\n",
    "    newdataframe.to_csv(\n",
    "        '../processeddata/simulations/runs678_stallstrengthfits_' + model +\n",
    "        '.tsv',\n",
    "        sep='\\t',\n",
    "        index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": true,
    "editable": true
   },
   "source": [
    "### Create mRNA sequence files for Run 6 to vary initiation rate in different models (for Fig. 3A)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true
   },
   "outputs": [],
   "source": [
    "mutation_locations = [{6: 'cta'}, ]\n",
    "\n",
    "yfpmutants = dict()\n",
    "for mutant in mutation_locations:\n",
    "    key = '_'.join(['yfp'] + [\n",
    "        codon + str(location) for location, codon in mutant.items()\n",
    "    ])\n",
    "    yfpmutants[key] = list(yfp0)\n",
    "    leucodon_number = 0\n",
    "    for position in range(0, len(yfp0), 3):\n",
    "        currentcodon = yfp0[position:position + 3]\n",
    "        # proceed only if the codon is a Leu codon (which are all CTG in yfp0)\n",
    "        if currentcodon not in ['CTG']:\n",
    "            continue\n",
    "        leucodon_number += 1\n",
    "        for location in mutant.keys():\n",
    "            if leucodon_number == location:\n",
    "                yfpmutants[key][position:position + 3] = mutant[\n",
    "                    location].upper()\n",
    "    yfpmutants[key] = ''.join(yfpmutants[key])\n",
    "\n",
    "defaultMrnaCopyNumber = 1  # per cell\n",
    "defaultInitationRate = 0.3  # s-1, This is the median initiation rate\n",
    "\n",
    "listOfInitiationRates = defaultInitationRate * 2**np.array(\n",
    "    range(-4, 4, 1), dtype=np.float64)\n",
    "\n",
    "for initiationRate in listOfInitiationRates:\n",
    "    for mutant in yfpmutants:\n",
    "        outputFile = '../annotations/simulations/run6/' + \\\n",
    "        '%s_initiationrate_%0.4g.csv'%(mutant, initiationRate)\n",
    "        num_seq = ''.join(\n",
    "            get_numerical_codon_sequence(yfpmutants[mutant][:-3]))\n",
    "        File = open(outputFile, 'w')\n",
    "\n",
    "        File.write(\"%0.4g\\t%d\\t%s\\n\" %\n",
    "                   (initiationRate, defaultMrnaCopyNumber,\n",
    "                    get_numerical_codon_sequence(yfp0[:-3])))\n",
    "\n",
    "        File.write(\"%0.4g\\t%d\\t%s\\n\" % (initiationRate, defaultMrnaCopyNumber,\n",
    "                                        num_seq))\n",
    "    File.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": true,
    "editable": true
   },
   "source": [
    "### Create mRNA sequence files for Run 7 to vary number of stall sites in different models (for Fig. 3B)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {
    "collapsed": true,
    "deletable": true,
    "editable": true
   },
   "outputs": [],
   "source": [
    "mutation_locations = [\n",
    "    {\n",
    "        2: 'cta'\n",
    "    },\n",
    "    {\n",
    "        6: 'cta'\n",
    "    },\n",
    "    {\n",
    "        10: 'cta'\n",
    "    },\n",
    "    {\n",
    "        14: 'cta'\n",
    "    },\n",
    "    {\n",
    "        18: 'cta'\n",
    "    },\n",
    "    {\n",
    "        2: 'cta',\n",
    "        6: 'cta'\n",
    "    },\n",
    "    {\n",
    "        2: 'cta',\n",
    "        6: 'cta',\n",
    "        10: 'cta'\n",
    "    },\n",
    "    {\n",
    "        2: 'cta',\n",
    "        6: 'cta',\n",
    "        10: 'cta',\n",
    "        14: 'cta'\n",
    "    },\n",
    "    {\n",
    "        2: 'cta',\n",
    "        6: 'cta',\n",
    "        10: 'cta',\n",
    "        14: 'cta',\n",
    "        18: 'cta'\n",
    "    },\n",
    "]\n",
    "\n",
    "yfpmutants = dict()\n",
    "for mutant in mutation_locations:\n",
    "    key = '_'.join(['yfp'] + [\n",
    "        codon + str(location) for location, codon in mutant.items()\n",
    "    ])\n",
    "    yfpmutants[key] = list(yfp0)\n",
    "    leucodon_number = 0\n",
    "    for position in range(0, len(yfp0), 3):\n",
    "        currentcodon = yfp0[position:position + 3]\n",
    "        # proceed only if the codon is a Leu codon (which are all CTG in yfp0)\n",
    "        if currentcodon not in ['CTG']:\n",
    "            continue\n",
    "        leucodon_number += 1\n",
    "        for location in mutant.keys():\n",
    "            if leucodon_number == location:\n",
    "                yfpmutants[key][position:position + 3] = mutant[\n",
    "                    location].upper()\n",
    "    yfpmutants[key] = ''.join(yfpmutants[key])\n",
    "\n",
    "defaultMrnaCopyNumber = 1  # per cell\n",
    "defaultInitationRate = 0.3  # s-1, This is the median initiation rate\n",
    "\n",
    "listOfInitiationRates = [defaultInitationRate]\n",
    "\n",
    "for initiationRate in listOfInitiationRates:\n",
    "    for mutant in yfpmutants:\n",
    "        outputFile = '../annotations/simulations/run7/' + \\\n",
    "        '%s_initiationrate_%0.4g.csv'%(mutant, initiationRate)\n",
    "        num_seq = ''.join(\n",
    "            get_numerical_codon_sequence(yfpmutants[mutant][:-3]))\n",
    "        File = open(outputFile, 'w')\n",
    "\n",
    "        File.write(\"%0.4g\\t%d\\t%s\\n\" %\n",
    "                   (initiationRate, defaultMrnaCopyNumber,\n",
    "                    get_numerical_codon_sequence(yfp0[:-3])))\n",
    "\n",
    "        File.write(\"%0.4g\\t%d\\t%s\\n\" % (initiationRate, defaultMrnaCopyNumber,\n",
    "                                        num_seq))\n",
    "    File.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": true,
    "editable": true
   },
   "source": [
    "### Create mRNA sequence files for Run 8 to vary distance between stall sites in different models (for Fig. 3C)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {
    "collapsed": true,
    "deletable": true,
    "editable": true
   },
   "outputs": [],
   "source": [
    "mutation_locations = [{loop: 'cta'} for loop in range(1, 238)]\n",
    "\n",
    "yfpmutants = dict()\n",
    "for mutant in mutation_locations:\n",
    "    key = '_'.join(['yfp'] + [\n",
    "        codon + str(location) for location, codon in mutant.items()\n",
    "    ])\n",
    "    yfpmutants[key] = list(yfp0)\n",
    "    codon_number = 0\n",
    "    for position in range(0, len(yfp0), 3):\n",
    "        currentcodon = yfp0[position:position + 3]\n",
    "        codon_number += 1\n",
    "        for location in mutant.keys():\n",
    "            if codon_number == location:\n",
    "                yfpmutants[key][position:position + 3] = mutant[\n",
    "                    location].upper()\n",
    "        yfpmutants\n",
    "        if position == 177:\n",
    "            yfpmutants[key][position:position + 3] = 'CTA'\n",
    "    yfpmutants[key] = ''.join(yfpmutants[key])\n",
    "\n",
    "defaultMrnaCopyNumber = 1  # per cell\n",
    "defaultInitationRate = 0.3  # s-1, This is the median initiation rate\n",
    "\n",
    "listOfInitiationRates = [defaultInitationRate]\n",
    "\n",
    "for initiationRate in listOfInitiationRates:\n",
    "    for mutant in yfpmutants:\n",
    "        outputFile = '../annotations/simulations/run8/' + \\\n",
    "        '%s_initiationrate_%0.4g.csv'%(mutant, initiationRate)\n",
    "        num_seq = ''.join(\n",
    "            get_numerical_codon_sequence(yfpmutants[mutant][:-3]))\n",
    "        File = open(outputFile, 'w')\n",
    "\n",
    "        File.write(\"%0.4g\\t%d\\t%s\\n\" %\n",
    "                   (initiationRate, defaultMrnaCopyNumber,\n",
    "                    get_numerical_codon_sequence(yfp0[:-3])))\n",
    "\n",
    "        File.write(\"%0.4g\\t%d\\t%s\\n\" % (initiationRate, defaultMrnaCopyNumber,\n",
    "                                        num_seq))\n",
    "    File.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Create stall strength file for Run 11 CTA distance mutant prediction based on fit of Run 2 to single mutant data (for Fig. 6)\n",
    "\n",
    "Also print elongation rate fit shown in Table S5"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "tRNA accommodation rate at CTA67 (CSAT model)\t0.0103s-1\n",
      "tRNA accommodation rate at CTA118 (CSAT model)\t0.0146s-1\n",
      "tRNA accommodation rate at CTA124 (CSAT model)\t0.0158s-1\n",
      "tRNA accommodation rate at CTA136 (CSAT model)\t0.0159s-1\n",
      "tRNA accommodation rate at CTA140 (CSAT model)\t0.0183s-1\n",
      "tRNA accommodation rate at CTA59 (CSAT model)\t0.0183s-1\n",
      "tRNA accommodation rate at CTA63 (CSAT model)\t0.012s-1\n",
      "tRNA accommodation rate at CTA67 (TJ model)\t0.00825s-1\n",
      "tRNA accommodation rate at CTA118 (TJ model)\t0.0112s-1\n",
      "tRNA accommodation rate at CTA124 (TJ model)\t0.0113s-1\n",
      "tRNA accommodation rate at CTA136 (TJ model)\t0.0116s-1\n",
      "tRNA accommodation rate at CTA140 (TJ model)\t0.0127s-1\n",
      "tRNA accommodation rate at CTA59 (TJ model)\t0.013s-1\n",
      "tRNA accommodation rate at CTA63 (TJ model)\t0.00923s-1\n",
      "tRNA accommodation rate at CTA67 (SAT model)\t0.0348s-1\n",
      "tRNA accommodation rate at CTA118 (SAT model)\t0.0463s-1\n",
      "tRNA accommodation rate at CTA124 (SAT model)\t0.0529s-1\n",
      "tRNA accommodation rate at CTA136 (SAT model)\t0.0517s-1\n",
      "tRNA accommodation rate at CTA140 (SAT model)\t0.0607s-1\n",
      "tRNA accommodation rate at CTA59 (SAT model)\t0.0672s-1\n",
      "tRNA accommodation rate at CTA63 (SAT model)\t0.0408s-1\n"
     ]
    }
   ],
   "source": [
    "fitdatafolder = '../processeddata/simulations/'\n",
    "fitdatafiles = os.listdir(fitdatafolder)\n",
    "fitdatafiles = [\n",
    "    fitdatafolder + File for File in fitdatafiles\n",
    "    if File.startswith('run2_fit_stallstrength_for_cta_distance_')\n",
    "]\n",
    "\n",
    "# find the location of all leucine codons to convert leu codon serial number\n",
    "# to absolute position along yfp in codon units for simulation\n",
    "leupositions = dict()\n",
    "leucodon_number = 1\n",
    "for position in range(0, len(yfp0), 3):\n",
    "    currentcodon = yfp0[position:position + 3]\n",
    "    if currentcodon == 'CTG':\n",
    "        leupositions[leucodon_number] = position / 3\n",
    "        leucodon_number += 1\n",
    "\n",
    "# extracted from simulation run 1\n",
    "trnaaccommodationrates = {23: 0.132824, 24: 0.398473, 26: 0.255022}\n",
    "codonnames = {23: 'CTA', 24: 'CTC', 26: 'CTT'}\n",
    "modelnames = {\n",
    "    'trafficjam': 'TJ',\n",
    "    'selpreterm': 'SAT',\n",
    "    '5primepreterm': 'CSAT',\n",
    "}\n",
    "\n",
    "for fitdatafile in fitdatafiles:\n",
    "    fitdata = pd.read_table(fitdatafile)\n",
    "    fitdata['codon'] = fitdata['mutant'].apply(lambda string: string[:3])\n",
    "    fitdata['pos'] = fitdata['mutant'].apply(lambda string: int(string[3:]))\n",
    "    fitdata = fitdata.drop(['mutant'], axis=1)\n",
    "    fitdata['pos'] = fitdata['pos'].apply(lambda pos: leupositions[pos])\n",
    "    fitdata['codon'] = fitdata['codon'].apply(\n",
    "        lambda codon: codonDict[codon.upper()])\n",
    "    fitdata = fitdata[['codon', 'pos', 'stallstrength']]\n",
    "    model = fitdatafile.split('_')[-1].split('.')[0]\n",
    "    fitdata.to_csv(\n",
    "        '../processeddata/simulations/run11_stallstrengthfits_' + model +\n",
    "        '.tsv',\n",
    "        sep='\\t',\n",
    "        index=False)\n",
    "    fitdata['trnaaccommodationrate'] = fitdata['codon'].apply(\n",
    "        lambda x: trnaaccommodationrates[x])\n",
    "    fitdata['trnaaccommodationrate'] = fitdata[\n",
    "        'trnaaccommodationrate'] * fitdata['stallstrength']\n",
    "    for row in fitdata.iterrows():\n",
    "        codon = row[1]['codon']\n",
    "        pos = row[1]['pos']\n",
    "        rate = row[1]['trnaaccommodationrate']\n",
    "        print 'tRNA accommodation rate at {0}{1} ({2} model)\\t{3:.3}s-1'.format(\n",
    "            codonnames[codon], int(pos), modelnames[model], rate)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Create mRNA sequence files for Run 11 to predict CTA distance mutant YFP synthesis rates in different models (for Fig. 6)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "mutation_locations = [\n",
    "    {\n",
    "        8: 'cta'\n",
    "    },\n",
    "    {\n",
    "        9: 'cta'\n",
    "    },\n",
    "    {\n",
    "        10: 'cta'\n",
    "    },\n",
    "    {\n",
    "        11: 'cta'\n",
    "    },\n",
    "    {\n",
    "        12: 'cta'\n",
    "    },\n",
    "    {\n",
    "        13: 'cta'\n",
    "    },\n",
    "    {\n",
    "        14: 'cta'\n",
    "    },\n",
    "    {\n",
    "        8: 'cta',\n",
    "        9: 'cta'\n",
    "    },\n",
    "    {\n",
    "        8: 'cta',\n",
    "        10: 'cta'\n",
    "    },\n",
    "    {\n",
    "        8: 'cta',\n",
    "        11: 'cta'\n",
    "    },\n",
    "    {\n",
    "        8: 'cta',\n",
    "        12: 'cta'\n",
    "    },\n",
    "    {\n",
    "        8: 'cta',\n",
    "        13: 'cta'\n",
    "    },\n",
    "    {\n",
    "        8: 'cta',\n",
    "        14: 'cta'\n",
    "    },\n",
    "    {\n",
    "        9: 'cta',\n",
    "        10: 'cta'\n",
    "    },\n",
    "    {\n",
    "        9: 'cta',\n",
    "        11: 'cta'\n",
    "    },\n",
    "    {\n",
    "        9: 'cta',\n",
    "        12: 'cta'\n",
    "    },\n",
    "    {\n",
    "        9: 'cta',\n",
    "        13: 'cta'\n",
    "    },\n",
    "    {\n",
    "        9: 'cta',\n",
    "        14: 'cta'\n",
    "    },\n",
    "    {\n",
    "        11: 'cta',\n",
    "        12: 'cta'\n",
    "    },\n",
    "    {\n",
    "        13: 'cta',\n",
    "        14: 'cta'\n",
    "    },\n",
    "]\n",
    "\n",
    "yfpmutants = dict()\n",
    "for mutant in mutation_locations:\n",
    "    key = '_'.join(['yfp'] + [\n",
    "        codon + str(location) for location, codon in mutant.items()\n",
    "    ])\n",
    "    yfpmutants[key] = list(yfp0)\n",
    "    leucodon_number = 0\n",
    "    for position in range(0, len(yfp0), 3):\n",
    "        currentcodon = yfp0[position:position + 3]\n",
    "        # proceed only if the codon is a Leu codon (which are all CTG in yfp0)\n",
    "        if currentcodon not in ['CTG']:\n",
    "            continue\n",
    "        leucodon_number += 1\n",
    "        for location in mutant.keys():\n",
    "            if leucodon_number == location:\n",
    "                yfpmutants[key][position:position + 3] = mutant[\n",
    "                    location].upper()\n",
    "    yfpmutants[key] = ''.join(yfpmutants[key])\n",
    "\n",
    "defaultMrnaCopyNumber = 1  # per cell\n",
    "defaultInitationRate = 0.3  # s-1, This is the median initiation rate\n",
    "\n",
    "listOfInitiationRates = [defaultInitationRate]\n",
    "\n",
    "for initiationRate in listOfInitiationRates:\n",
    "    for mutant in yfpmutants:\n",
    "        outputFile = '../annotations/simulations/run11/' + \\\n",
    "        '%s_initiationrate_%0.4g.csv'%(mutant, initiationRate)\n",
    "        num_seq = ''.join(\n",
    "            get_numerical_codon_sequence(yfpmutants[mutant][:-3]))\n",
    "        File = open(outputFile, 'w')\n",
    "\n",
    "        File.write(\"%0.4g\\t%d\\t%s\\n\" %\n",
    "                   (initiationRate, defaultMrnaCopyNumber,\n",
    "                    get_numerical_codon_sequence(yfp0[:-3])))\n",
    "\n",
    "        File.write(\"%0.4g\\t%d\\t%s\\n\" % (initiationRate, defaultMrnaCopyNumber,\n",
    "                                        num_seq))\n",
    "    File.close()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
